src/backend/joanie/core/utils/contract.py

```diff
@@ -0,0 +1,73 @@
+"""Utility to generate a zipfile of PDF bytes files for contracts that are signed"""
+import io
+import zipfile
+
+from django.core.files.storage import default_storage
+from django.db.models import Q
+
+from joanie.core import enums, models
+from joanie.signature.backends import get_signature_backend
+
+
+def get_signature_backend_references(
+    course_product_relation: models.CourseProductRelation,
+):
+    """
+    Get a list of signature backend references from a course product relation object that are
+    attached to a validated orders if it exists, else it returns an empty string.
+    """
+    signature_backend_references = (
+        models.Order.objects.filter(
+            Q(course_id=course_product_relation.course.pk)
+            | Q(product_id=course_product_relation.product.pk),
+            contract__signed_on__isnull=False,
+            state=enums.ORDER_STATE_VALIDATED,
+        )
+        .select_related("contract")
+        .values_list("contract__signature_backend_reference", flat=True)
+        .order_by("contract__signature_backend_reference")
+    )
+    return list(signature_backend_references)
+
+
+def retrieve_pdf_bytes_of_contracts(
+    signature_backend_references: list,
+):
+    """
+    Retrieve PDF bytes files from a list of signature backend reference at the signature provider.
+    """
+    signature_backend = get_signature_backend()
+
+    pdf_bytes_collection = [
+        signature_backend.get_signed_file(reference_id)
+        for reference_id in signature_backend_references
+    ]
+
+    return pdf_bytes_collection
+
+
+def generate_zipfile(pdf_bytes_list: list):
+    """
+    Generate a zipfile from a list of PDF bytes and save the ZIP archive in
+    default storage. The method returns the filename of the ZIP archive once done. This filename
+    can be used to retrieve it from default storage.
+
+    Used compression method `zipfile.ZIP_DEFLATED`, It is efficient in terms of compression and
+    decompression, making it a good choice for general-purpose compression.
+    """
+    # Create the ZIP Archive.
+    zip_buffer = io.BytesIO()
+    with zipfile.ZipFile(zip_buffer, "w", zipfile.ZIP_DEFLATED) as zipf:
+        # Iterate through each pdf bytes in the list.
+        for index, pdf_bytes in enumerate(pdf_bytes_list):
+            pdf_filename = f"contract_{index}.pdf"
+            # Add PDF bytes file in ZIP archive.
+            zipf.writestr(pdf_filename, pdf_bytes)
+
```

Comment:
No need to `close` zip file ?

Reply:
When we use the "with" keyword, it gets closed when we get out of it. If I don't use "with" keyword, then eventually I will need to close it manually.